src/main/java/frc/robot/systems/IntakeFSMSystem.java

```diff
@@ -0,0 +1,707 @@
+	 * and the current state of this FSM. This method should not have any side
+	 * effects on outputs. In other words, this method should only read or get
+	 * values to decide what state to go to.
+	 * @param input Global TeleopInput if robot in teleop mode or null if
+	 *        the robot is in autonomous mode.
+	 * @return FSM state for the next iteration
+	 */
+	private IntakeFSMState nextState(TeleopInput input) {
+		switch (currentState) {
+			case MOVE_TO_HOME:
+				if (input == null) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				}
+
+				if (input.isIntakeButtonPressed() || input.isOuttakeButtonPressed()) {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				} else {
+					if (input.isShootButtonPressed()) {
+						return IntakeFSMState.FEED_SHOOTER;
+					} else {
+						return IntakeFSMState.MOVE_TO_HOME;
+					}
+				}
+
+			case FEED_SHOOTER:
+				if (input.isShootButtonPressed() && !(input.isIntakeButtonPressed()
+					|| input.isOuttakeButtonPressed())) {
+					return IntakeFSMState.FEED_SHOOTER;
+				} else {
+					return IntakeFSMState.MOVE_TO_HOME;
+				}
+
+			case MOVE_TO_GROUND:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)) {
+					if (input.isIntakeButtonPressed() && !input.isOuttakeButtonPressed()
+						&& !input.isShootButtonPressed()) {
+						return IntakeFSMState.INTAKING;
+					} else if (input.isOuttakeButtonPressed() && !input.isIntakeButtonPressed()
+								&& !input.isShootButtonPressed()) {
+						return IntakeFSMState.OUTTAKING;
+					}
+				}
+
+				if ((input.isIntakeButtonPressed() || input.isOuttakeButtonPressed())
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				} else {
+					return IntakeFSMState.MOVE_TO_HOME;
+				}
+
+			case INTAKING:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)
+					&& input.isIntakeButtonPressed() && !input.isOuttakeButtonPressed()
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.INTAKING;
+				}
+
+				if (input.isShootButtonPressed()
+					&& !(input.isOuttakeButtonPressed() || input.isIntakeButtonPressed())) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			case OUTTAKING:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)
+					&& input.isOuttakeButtonPressed() && !input.isIntakeButtonPressed()
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.OUTTAKING;
+				}
+
+				if (input.isShootButtonPressed()
+					&& !(input.isOuttakeButtonPressed() || input.isIntakeButtonPressed())) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			default:
+				throw new IllegalStateException("Invalid state: " + currentState.toString());
+		}
+	}
+
+/* ------------------------ Command handlers ------------------------ */
+
+	private boolean inRange(double targetEncoder, double currentEncoder) {
```

Comment:
I feel like a more appropriate names would along the lines of "approximatelyEquals"